File: components/power.py
import os
import subprocess
from PyQt6.QtWidgets import (
    QLabel, QWidget, QVBoxLayout, QHBoxLayout, QPushButton,
    QFrame, QStackedLayout
)
from PyQt6.QtGui import QPixmap, QFont
from PyQt6.QtCore import Qt


class PowerManager:
    def __init__(self, parent):
        self.parent = parent
        self.selected_mode = None
        self.mode_boxes = {}

        self.init_ui()

    def init_ui(self):
        self.power_box = QWidget(self.parent)
        self.power_box.setGeometry(455, 100, 700, 470)
        self.power_box.setStyleSheet("""
            background-color: rgba(255, 255, 255, 0.05);
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
        """)
        self.power_box.setVisible(False)

        layout = QVBoxLayout(self.power_box)
        layout.setContentsMargins(150, 100, 150, 200)
        layout.setSpacing(25)

        power_label = QLabel("POWER options", self.power_box)
        power_label.setFont(QFont("ElysiaOSNew", 16))
        power_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
        power_label.setGeometry(270, 0, 500, 160)

        mode_widget = QWidget(self.power_box)
        mode_widget.setGeometry(110, 320, 110, 250)
        mode_container = QHBoxLayout(mode_widget)
        layout.addWidget(mode_widget)

        modes = {
            "performance": ("Performance", "performance.png"),
            "balanced": ("Balanced", "balanced.png"),
            "default": ("Default", "default.png")
        }

        for mode_key, (label_text, image_file) in modes.items():
            outer_layout = QVBoxLayout()
            outer_layout.setAlignment(Qt.AlignmentFlag.AlignCenter)
            outer_layout.setSpacing(8)

            label = QLabel(label_text)
            label.setFont(QFont("ElysiaOSNew", 12))
            label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            label.setAlignment(Qt.AlignmentFlag.AlignCenter)

            frame = QFrame(self.power_box)
            frame.setFixedSize(130, 130)
            frame.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            frame.setCursor(Qt.CursorShape.PointingHandCursor)

            stacked_layout = QStackedLayout(frame)

            image_label = QLabel()
            image_path = os.path.expanduser(f"~/Elysia/assets/assets/{image_file}")
            if os.path.exists(image_path):
                pixmap = QPixmap(image_path).scaled(130, 130, Qt.AspectRatioMode.KeepAspectRatio, Qt.TransformationMode.SmoothTransformation)
                image_label.setPixmap(pixmap)
            else:
                image_label.setText("Missing")

            tick = QLabel("✔")
            tick.setStyleSheet("color: #e5a7c6; font-size: 24px; background: transparent; border: none;")
            tick.setAlignment(Qt.AlignmentFlag.AlignTop | Qt.AlignmentFlag.AlignRight)
            tick.setVisible(False)

            stacked_layout.addWidget(image_label)
            stacked_layout.addWidget(tick)

            frame.mousePressEvent = lambda _, m=mode_key: self.select_mode(m)
            self.mode_boxes[mode_key] = (frame, tick)

            outer_layout.addWidget(label)
            outer_layout.addWidget(frame)

            mode_container.addLayout(outer_layout)

        layout.addLayout(mode_container)

        self.apply_btn = QPushButton("Apply Changes", self.power_box)
        self.apply_btn.setFixedSize(200, 50)
        self.apply_btn.setStyleSheet("""
            QPushButton {
                background-color: #e5a7c6;
                color: white;
                font-weight: bold;
                border: none;
                padding: 6px 12px;
                letter-spacing: 1px;
                border-radius: 10px;
                font-size: 16px;
            }
            QPushButton:hover {
                background-color: #edcee3;
            }
        """)
        self.apply_btn.clicked.connect(self.apply_changes)
        self.apply_btn.setGeometry(250, 370, 200, 150)

    def show(self):
        self.parent.hide_tiles()
        self.parent.show_back_button()
        self.power_box.setVisible(True)
        self.auto_detect_current_mode()

    def hide(self):
        self.power_box.setVisible(False)
        self.parent.hide_back_button()
        self.parent.show_tiles()

    def select_mode(self, mode):
        self.selected_mode = mode
        for name, (frame, tick) in self.mode_boxes.items():
            is_selected = name == mode
            frame.setStyleSheet(f"border: {'3px solid #e5a7c6' if is_selected else '2px solid transparent'}; border-radius: 26px; background: transparent;")
            tick.setVisible(is_selected)

    def apply_changes(self):
        if not self.selected_mode:
            return

        command = {
            "performance": "pkexec auto-cpufreq --force=performance",
            "balanced": "pkexec auto-cpufreq --force=powersave",
            "default": "pkexec auto-cpufreq --force=reset"
        }.get(self.selected_mode)

        if command:
            subprocess.Popen(command.split())

    def auto_detect_current_mode(self):
        try:
            output = subprocess.check_output(["pkexec", "auto-cpufreq", "--debug"], stderr=subprocess.STDOUT, text=True, timeout=15)
            if "Currently using: performance governor" in output:
                self.select_mode("performance")
            elif "Currently using: powersave governor" in output:
                self.select_mode("balanced")
            elif "Currently using: reset governor" in output:
                self.select_mode("default")
        except subprocess.SubprocessError:
            pass

File: components/Power.cpp
#include "Power.h"
#include "../translations.h"
#include <QDir>
#include <QFont>
#include <QProcess>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QFrame>
#include <QStackedLayout>
#include <QMap>
#include <QTimer>
#include <QFileInfo>
#include <QPixmap>
#include <QMessageBox>
#include <QtConcurrent/QtConcurrent>
#include <QFutureWatcher>

Power::Power(QObject *parent) : QObject(parent) {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent);
    if (!parentWidget) return;
    initUI();
}

void Power::initUI() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    powerBox = new QWidget(parentWidget);
    powerBox->setGeometry(455, 100, 700, 470);
    powerBox->setStyleSheet(
        "background-color: rgba(255, 255, 255, 0.05);"
        "border: 2px solid rgba(255, 255, 255, 0.15);"
        "border-radius: 30px;"
    );
    powerBox->setVisible(false);
    QVBoxLayout *layout = new QVBoxLayout(powerBox);
    layout->setContentsMargins(150, 100, 150, 200); // match Python
    layout->setSpacing(25);
    QLabel *powerLabel = new QLabel(Translations::tr("POWER_OPTIONS"), powerBox);
    powerLabel->setFont(QFont("ElysiaOSNew", 16));
    powerLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    powerLabel->setAlignment(Qt::AlignCenter);
    layout->addWidget(powerLabel);
    // Mode row as layout
    QHBoxLayout *modeContainer = new QHBoxLayout();
    modeContainer->setSpacing(30);
    modeContainer->setAlignment(Qt::AlignCenter);
    QMap<QString, QPair<QString, QString>> modes = {
        {"performance", {Translations::tr("PERFORMANCE"), "performance.png"}},
        {"balanced", {Translations::tr("BALANCED"), "balanced.png"}},
        {"default", {Translations::tr("DEFAULT"), "default.png"}}
    };
    for (auto it = modes.begin(); it != modes.end(); ++it) {
        QString modeKey = it.key();
        QString labelText = it.value().first;
        QString imageFile = it.value().second;
        QVBoxLayout *outerLayout = new QVBoxLayout();
        outerLayout->setAlignment(Qt::AlignCenter);
        outerLayout->setSpacing(8);
        // Label above icon
        QLabel *label = new QLabel(labelText);
        label->setFont(QFont("ElysiaOSNew", 12));
        label->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        label->setAlignment(Qt::AlignCenter);
        outerLayout->addWidget(label, 0, Qt::AlignCenter);
        // Mode tile frame
        QFrame *frame = new QFrame(powerBox);
        frame->setFixedSize(130, 130);
        frame->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        frame->setCursor(Qt::PointingHandCursor);
        QStackedLayout *stackedLayout = new QStackedLayout(frame);
        QLabel *imageLabel = new QLabel();
        QString iconPath = QDir::homePath() + "/.config/Elysia/assets/assets/" + imageFile;
        if (QFileInfo::exists(iconPath)) {
            QPixmap pixmap(iconPath);
            imageLabel->setPixmap(pixmap.scaled(130, 130, Qt::KeepAspectRatio, Qt::SmoothTransformation));
        } else {
            imageLabel->setText(Translations::tr("MISSING"));
        }
        QLabel *tick = new QLabel(QString::fromUtf8("\u2714"));
        tick->setStyleSheet("color: #e5a7c6; font-size: 24px; background: transparent; border: none;");
        tick->setAlignment(Qt::AlignTop | Qt::AlignRight);
        tick->setVisible(false);
        stackedLayout->addWidget(imageLabel);
        stackedLayout->addWidget(tick);
        frame->installEventFilter(this);
        frame->setProperty("modeKey", modeKey);
        modeBoxes[modeKey] = qMakePair(frame, tick);
        outerLayout->addWidget(frame, 0, Qt::AlignCenter);
        modeContainer->addLayout(outerLayout);
    }
    layout->addLayout(modeContainer);
    // Apply button
    applyBtn = new QPushButton(Translations::tr("APPLY_CHANGES"), powerBox);
    applyBtn->setFixedSize(170, 40);
    applyBtn->setStyleSheet(
        "QPushButton {"
        "    background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #e5a7c6, stop:1 #edcee3);"
        "    color: #fff;"
        "    font-weight: bold;"
        "    border: none;"
        "    padding: 6px 12px;"
        "    letter-spacing: 1px;"
        "    border-radius: 10px;"
        "    font-size: 15px;"
        "}"
        "QPushButton:hover {"
        "    background: #edcee3; color: #a87c4f;"
        "}"
    );
    applyBtn->setToolTip(Translations::tr("APPLY_TOOLTIP"));
    connect(applyBtn, &QPushButton::clicked, this, &Power::applyChanges);
    // Place button near the bottom center of the box
    applyBtn->move((680-140)/2, 400);
}

void Power::show() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("hideTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideTiles");
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("showBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showBackButton");
    powerBox->setVisible(true);
    // Run detection in background
    if (!modeDetectWatcher) {
        modeDetectWatcher = new QFutureWatcher<QString>(this);
        connect(modeDetectWatcher, &QFutureWatcher<QString>::finished, this, &Power::onModeDetectFinished);
    }
    QFuture<QString> future = QtConcurrent::run([=]() {
        QProcess process;
        QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
        env.insert("DISPLAY", qgetenv("DISPLAY"));
        env.insert("XAUTHORITY", qgetenv("XAUTHORITY"));
        env.insert("DBUS_SESSION_BUS_ADDRESS", qgetenv("DBUS_SESSION_BUS_ADDRESS"));
        process.setProcessEnvironment(env);
        process.start("pkexec", QStringList() << "auto-cpufreq" << "--debug");
        if (!process.waitForStarted(5000)) return QString();
        process.waitForFinished(15000);
        QString output = QString::fromUtf8(process.readAllStandardOutput());
        if (output.contains("Currently using: performance governor")) return QString("performance");
        if (output.contains("Currently using: powersave governor")) return QString("balanced");
        if (output.contains("Currently using: reset governor")) return QString("default");
        return QString();
    });
    modeDetectWatcher->setFuture(future);
}

void Power::onModeDetectFinished() {
    if (!modeDetectWatcher) return;
    QString mode = modeDetectWatcher->result();
    if (!mode.isEmpty()) selectMode(mode);
}

void Power::hide() {
    powerBox->setVisible(false);
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("hideBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideBackButton");
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("showTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showTiles");
}

void Power::selectMode(const QString &mode) {
    selectedMode = mode;
    for (auto it = modeBoxes.begin(); it != modeBoxes.end(); ++it) {
        bool isSelected = (it.key() == mode);
        it.value().first->setStyleSheet(QString("border: %1 solid #e5a7c6; border-radius: 26px; background: transparent;").arg(isSelected ? "3px" : "2px transparent"));
        it.value().second->setVisible(isSelected);
    }
}

void Power::applyChanges() {
    if (selectedMode.isEmpty()) return;
    QMap<QString, QStringList> commands = {
        {"performance", {"auto-cpufreq", "--force=performance"}},
        {"balanced", {"auto-cpufreq", "--force=powersave"}},
        {"default", {"auto-cpufreq", "--force=reset"}}
    };
    QStringList args = commands.value(selectedMode);
    if (!args.isEmpty()) {
        QProcess process;
        QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
        env.insert("DISPLAY", qgetenv("DISPLAY"));
        env.insert("XAUTHORITY", qgetenv("XAUTHORITY"));
        env.insert("DBUS_SESSION_BUS_ADDRESS", qgetenv("DBUS_SESSION_BUS_ADDRESS"));
        process.setProcessEnvironment(env);
        process.start("pkexec", args);
        if (!process.waitForStarted(5000)) {
            QMessageBox::warning(powerBox, Translations::tr("ERROR_FASTFETCH"), "Failed to start pkexec.\nCheck your polkit setup and environment.");
            return;
        }
        process.waitForFinished(-1); // Wait for user to complete polkit prompt
        if (process.exitStatus() != QProcess::NormalExit || process.exitCode() != 0) {
            QMessageBox::warning(powerBox, Translations::tr("ERROR_FASTFETCH"), "Failed to apply CPU profile.\n\nOutput:\n" + process.readAllStandardError());
        }
    }
}

void Power::autoDetectCurrentMode() {
    QProcess process;
    process.start("pkexec", QStringList() << "auto-cpufreq" << "--debug");
    process.waitForFinished(15000);
    QString output = QString::fromUtf8(process.readAllStandardOutput());
    if (output.contains("Currently using: performance governor")) {
        selectMode("performance");
    } else if (output.contains("Currently using: powersave governor")) {
        selectMode("balanced");
    } else if (output.contains("Currently using: reset governor")) {
        selectMode("default");
    }
}

// Event filter for mode frame clicks
#include <QEvent>
bool Power::eventFilter(QObject *obj, QEvent *event) {
    if (event->type() == QEvent::MouseButtonPress) {
        for (auto it = modeBoxes.begin(); it != modeBoxes.end(); ++it) {
            if (it.value().first == obj) {
                selectMode(it.key());
                break;
            }
        }
    }
    return QObject::eventFilter(obj, event);
} 